hcl/hcl_test.go

```diff
@@ -281,6 +342,225 @@ terramate {
 				err: hcl.ErrMalformedTerramateConfig,
 			},
 		},
+		{
+			name: "empty backend",
+			input: `
+	terramate {
+		   backend "something" {
+		   }
+	}
+	`,
+			want: want{
+				config: hcl.Config{
+					Terramate: &hcl.Terramate{
+						Backend: &hclsyntax.Block{
+							Type:   "backend",
+							Labels: []string{"something"},
+						},
+					},
+				},
+			},
+		},
+	} {
+		testParser(t, tc)
+	}
+}
+
+func TestHCLParserRootConfig(t *testing.T) {
+	for _, tc := range []testcase{
+		{
+			name: "no config returns empty config",
+			input: `
+terramate {
+
+}
+`,
+			want: want{
+				config: hcl.Config{
+					Terramate: &hcl.Terramate{},
+				},
+			},
+		},
+		{
+			name: "empty config block returns empty config",
+			input: `
+terramate {
+	config {}
+}
+`,
+			want: want{
+				config: hcl.Config{
+					Terramate: &hcl.Terramate{
+						RootConfig: &hcl.RootConfig{},
+					},
+				},
+			},
+		},
+		{
+			name: "unrecognized config attribute",
+			input: `
+terramate {
+	config {
+		something = "bleh"
+	}
+}
+`,
+			want: want{
+				err: hcl.ErrMalformedTerramateConfig,
+			},
+		},
+		{
+			name: "empty config.git block",
+			input: `
+terramate {
+	config {
+		git {}
+	}
+}
+`,
+			want: want{
+				config: hcl.Config{
+					Terramate: &hcl.Terramate{
+						RootConfig: &hcl.RootConfig{},
+					},
+				},
+			},
+		},
+		{
+			name: "multiple config blocks - fails",
+			input: `
+terramate {
+	config {}
+	config {}
+}
+	`,
+			want: want{
+				err: hcl.ErrMalformedTerramateConfig,
+			},
+		},
+		{
+			name: "basic config.git block",
+			input: `
+terramate {
+	config {
+		git {
+			branch = "trunk"
+		}
+	}
+}
+`,
+			want: want{
+				config: hcl.Config{
+					Terramate: &hcl.Terramate{
+						RootConfig: &hcl.RootConfig{
+							Git: hcl.GitConfig{
+								Branch: "trunk",
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "all fields set for config.git",
+			input: `
+terramate {
+	config {
+		git {
+			branch = "trunk"
+			remote = "upstream"
+			baseRef = "upstream/trunk"
+			defaultBranchBaseRef = "HEAD~2"
+		}
+	}
+}
+`,
+			want: want{
+				config: hcl.Config{
+					Terramate: &hcl.Terramate{
+						RootConfig: &hcl.RootConfig{
+							Git: hcl.GitConfig{
+								Branch:               "trunk",
+								Remote:               "upstream",
+								BaseRef:              "upstream/trunk",
+								DefaultBranchBaseRef: "HEAD~2",
+							},
+						},
+					},
+				},
+			},
+		},
+	} {
+		testParser(t, tc)
+	}
+}
+
+func TestHCLParserStack(t *testing.T) {
+	for _, tc := range []testcase{
+		{
+			name: "empty stack block",
+			input: `
+terramate {
+	required_version = ""
+}
+stack {}`,
+			want: want{
+				config: hcl.Config{
+					Terramate: &hcl.Terramate{},
+					Stack:     &hcl.Stack{},
+				},
+			},
+		},
+		{
+			name: "multiple stack blocks",
+			input: `stack{}
+stack {}`,
```

Comment:
On this input there is not terramate block, is this OK/expected ? (wondering if terramate is always required, or not, when, etc).

Reply:
yeah good catch, in this case it's failing for multiple stacks, but if there was only 1 declaration then it'd fail with ErrNoTerramateBlock. Will fix.